libraries/AR_WPNav/AR_WPNav.cpp

```diff
@@ -0,0 +1,371 @@
+    // @Description: Waypoint speed default
+    // @Units: m/s
+    // @Range: 0 100
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("SPEED", 1, AR_WPNav, _speed_max, AR_WPNAV_SPEED_DEFAULT),
+
+    // @Param: RADIUS
+    // @DisplayName: Waypoint radius
+    // @Description: The distance in meters from a waypoint when we consider the waypoint has been reached. This determines when the vehicle will turn toward the next waypoint.
+    // @Units: m
+    // @Range: 0 100
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("RADIUS", 2, AR_WPNav, _radius, AR_WPNAV_RADIUS_DEFAULT),
+
+    // @Param: OVERSHOOT
+    // @DisplayName: Waypoint overshoot maximum
+    // @Description: Waypoint overshoot maximum in meters.  The vehicle will attempt to stay within this many meters of the track as it completes one waypoint and moves to the next.
+    // @Units: m
+    // @Range: 0 10
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("OVERSHOOT", 3, AR_WPNav, _overshoot, AR_WPNAV_OVERSHOOT_DEFAULT),
+
+    // @Param: PIVOT_ANGLE
+    // @DisplayName: Waypoint Pivot Angle
+    // @Description: Pivot when the difference bewteen the vehicle's heading and it's target heading is more than this many degrees.  Set to zero to disable pivot turns
+    // @Units: deg
+    // @Range: 0 360
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("PIVOT_ANGLE", 4, AR_WPNav, _pivot_angle, AR_WPNAV_PIVOT_ANGLE_DEFAULT),
+
+    // @Param: PIVOT_RATE
+    // @DisplayName: Waypoint Pivot Turn Rate
+    // @Description: Turn rate during pivot turns
+    // @Units: deg/s
+    // @Range: 0 360
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("PIVOT_RATE", 5, AR_WPNav, _pivot_rate, AR_WPNAV_PIVOT_RATE_DEFAULT),
+
+    AP_GROUPEND
+};
+
+AR_WPNav::AR_WPNav(AR_AttitudeControl& atc, AP_Navigation& nav_controller) :
+    _atc(atc),
+    _nav_controller(nav_controller)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+// update navigation
+void AR_WPNav::update(float dt)
+{
+    // exit immediately if no current location, origin or destination
+    Location current_loc;
+    float speed;
+    if (!hal.util->get_soft_armed() || !_orig_and_dest_valid || !AP::ahrs().get_position(current_loc) || !_atc.get_forward_speed(speed)) {
+        _desired_speed_limited = _atc.get_desired_speed_accel_limited(0.0f, dt);
+        _desired_turn_rate_rads = 0.0f;
+        return;
+    }
+
+    // if no recent calls initialise desired_speed_limited to current speed
+    if (!is_active()) {
+        _desired_speed_limited = speed;
+    }
+    _last_update_ms = AP_HAL::millis();
+
+    update_distance_and_bearing_to_destination();
+
+    // check if vehicle has reached the destination
+    const bool near_wp = _distance_to_destination <= _radius;
+    const bool past_wp = current_loc.past_interval_finish_line(_origin, _destination);
+    if (!_reached_destination && (near_wp || past_wp)) {
```

Comment:
I was given a hand slap for doing a single assignment inside an if 😄 I like them but it would look over complicated here I think.

Reply:
Not sure I totally understand.. but I guess I could move the (near_wp || past_wp) one line lower.  We need to be careful that _reached_destination doesn't become false because of this line..